src/tls/client-certificate.c

```diff
@@ -0,0 +1,419 @@
+{
+  ssize_t s = write (fd, data, size);
+
+  if (s == size)
+    return true;
+
+  if (s == -1)
+    warn ("Couldn't write %s to certificate file", description);
+  else
+    warnx ("Partial write of %s to certificate file: %zu of %zu", description, s, size);
+
+  return false;
+}
+
+static bool
+client_certificate_write_cgroup_header (int         fd,
+                                        const char *wsinstance)
+{
+  /* First write the expected cgroup of the wsinstance */
+  char header[200];
+  int s = snprintf (header, sizeof header,
+                    "0::/system.slice/system-cockpithttps.slice/cockpit-wsinstance-https@%s.service\n",
+                    wsinstance);
+  assert (s < sizeof header);
+
+  return client_certificate_write_data (fd, header, s, "cgroup header");
+}
+
+static bool
+client_certificate_write_pem (int   fd,
+                              const gnutls_datum_t *der)
+{
+  gnutls_datum_t pem = { NULL, 0 };
+  int r = gnutls_pem_base64_encode2 ("CERTIFICATE", der, &pem);
+  if (r != GNUTLS_E_SUCCESS)
+    {
+      warnx ("Couldn't base64 encode certificate: %s", gnutls_strerror (r));
+      return false;
+    }
+
+  bool result = client_certificate_write_data (fd, pem.data, pem.size, "PEM data");
+
+  /* Make sure we get the function version and not the weird
+   * side-effecting macro version.
+   */
+  (gnutls_free) (pem.data);
+
+  return result;
+}
+
+/**
+ * client_certificate_link_fd_to_random_name:
+ * @dirfd: the directory to link the file
+ * @fd: the file created with O_TMPFILE
+ * @out_filename: the filename that was chosen
+ *
+ * Links the O_TMPFILE referred to by @fd to a random filename in
+ * @dirfd.
+ *
+ * On success, %true is returned and @out_filename is set to the
+ * filename that was used.  It must be free()d.
+ *
+ * On failure, %false is returned and a message will have been logged.
+ */
+static bool
+client_certificate_link_fd_to_random_name (int    dirfd,
+                                           int    fd,
+                                           char **out_filename)
+{
+  char *filename = client_certificate_random_filename ();
+
+  /* "the usual tricks" — see openat(2) and linkat(2) */
+  char path[PATH_MAX];
+  snprintf(path, PATH_MAX,  "/proc/self/fd/%d", fd);
+
+  if (linkat (AT_FDCWD, path, dirfd, filename, AT_SYMLINK_FOLLOW) == 0)
+    {
+      *out_filename = filename;
+      return true;
+    }
+  else
+    {
+      warn ("Unable to link client certificate file to /run/cockpit/tls/%s", filename);
+      free (filename);
+      return false;
+    }
+}
+
+/**
+ * client_certificate_create_tmpfile:
+ * @dirfd: the directory that will eventually contain the file
+ * @out_fd: a write-mode fd for the created file
+ */
+static bool
+client_certificate_create_tmpfile (int  dirfd,
+                                   int *out_fd)
+{
+  /* create with no permissions: only root will read this */
+  int fd = openat (dirfd, ".", O_TMPFILE | O_WRONLY, 0);
```

Comment:
Still, what's wrong with 0400, so that cockpit-tls itself can at least read it? Doesn't 0 cause troubles with unit tests?

Reply:
I changed this to `0400`.  It turned out to be necessary for testing.